sd-boot,bootctl,gpt-auto: support Extended Boot Loader Partition

poettering opened the pull request:
This is a larger PR that adds full support to the Extended Boot Loader partition (herein dubbed "XBOOTLDR") that is defined in https://systemd.io/BOOT_LOADER_SPECIFICATION for a long time, but never actually implemented in our codebase. Support is added to all layers:

* sd-boot (we'll now search for XBOOTLDR partitions, and honour the drop-ins and unified kernels in them)
* bootctl (we'll now list entries from both the ESP and XBOOTLDR)
* systemd-gpt-auto-generator (we'll now mount XBOOTLDR to /boot if it exists)

This also adds support for Type 2 boot loader snippets to bootctl/bootspec.c.

With all this in place we should finally implement 100% of what the boot loader specification defines.

Supporting XBOOTLDR hopefully makes sd-boot a bit more digestable for generic distros which do not want to resize an existing ESP at install time, since it means they can now place kernels in a separate XBOOTLDR partition, and only install sd-boot itself in the ESP.

poettering commented:
This goes nicely with https://github.com/systemd/mkosi/pull/326 btw.

kirbyfan64 commented:
To be clear, this would essentially allow boot information to be somewhere other than the ESP? Would the filesystem be any supported by the UEFI system drivers?

poettering commented:
> To be clear, this would essentially allow boot information to be somewhere other than the ESP? 

Well, yes, but that "somewhere" is only the XBOOTLDR partition and you can only have one of those per disk (well, strictly speaking you could have more, but this code only checks the first one, much like for the ESP)

> Would the filesystem be any supported by the UEFI system drivers?

Yes, this code makes no restriction on the file system used (which is different from the handling of the ESP where we explicitly check for FAT). In fact in a future patch I'd like to add a special directory in the ESP to sd-boot where you can drop EFI drivers into, and sd-boot will load them all. The idea is then that people can drop weird fs drivers in there, and sd-boot will load them and then have them available when searching for XBOOTLDR. That way we can support other fs, in a fashion of very lose coupling, as sd-boot needs to know nothing about fs drivers, it just loads everything dropped in.

keszybz commented on src/shared/bootspec.c:
```diff
@@ -265,6 +269,245 @@ static int boot_entries_find(
+                        continue;
+
+                if (*b)
+                        return log_error_errno(SYNTHETIC_ERRNO(EBADMSG), "Duplication section %s, refusing.", sections[i].Name);
```
"Duplicated"

keszybz commented on src/shared/bootspec.c:
```diff
@@ -265,6 +269,245 @@ static int boot_entries_find(
+                           "VERSION_ID", &version_id,
+                           "BUILD_ID", &build_id);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse os-release buffer: %m");
```
Hmm, I think it'd much more reasonable to indicate to the user where this data is coming from. I'd go even as far as passing the path to the file and using it here. The fact that we're using a buffer is an internal implementation detail.

keszybz commented on src/shared/bootspec.c:
```diff
@@ -265,6 +269,245 @@ static int boot_entries_find(
+                return log_error_errno(r, "Failed to parse os-release buffer: %m");
+
+        if (!os_pretty_name || !os_id || !(version_id || build_id))
+                return log_error_errno(SYNTHETIC_ERRNO(EBADMSG), "Missing OS release fields, refusing.");
```
Here too. Right now the user might be confused what is there to fix.

keszybz commented on src/shared/bootspec.c:
```diff
@@ -392,13 +635,23 @@ int boot_entries_load_config(
+                p = strjoina(esp_path, "/EFI/Linux/");
+                r = boot_entries_find_unified(p, esp_path, &config->entries, &config->n_entries);
+                if (r < 0)
+                        return r;
```
Hmm, if any of the entries are invalid (e.g. missing os-release fields), we're fail hard here. I guess it'd be better to warn and ignore invalid entries.

keszybz commented on src/shared/bootspec.c:
```diff
@@ -654,6 +644,8 @@ int boot_entries_load_config(
                         return r;
         }
 
+        typesafe_qsort(config->entries, config->n_entries, boot_entry_compare);
```
So much simpler and more efficient at the same time. Progress!

keszybz commented on src/boot/efi/boot.c:
```diff
@@ -2142,7 +2142,7 @@ EFI_STATUS efi_main(EFI_HANDLE image, EFI_SYSTEM_TABLE *sys_table) {
-        config_entry_add_linux(&config, loaded_image, root_dir);
+        config_entry_add_linux(&config, loaded_image->DeviceHandle, root_dir);
 
         /* scan /loader/entries/\*.conf files */
```
Stray backslash?

poettering replied:
i guess, it has been in the sources for about forever. will sneak a fix in

poettering replied:
hmm, nope, it's on purpose, otherwise gcc complains that /* is contained within a comment...

keszybz commented on src/boot/efi/boot.c:
```diff
@@ -1955,6 +1957,191 @@ static VOID config_entry_add_linux(
+                        continue;
+
+                /* Filter out some block devices early. */
+                if (block_io->Media->LogicalPartition ||
```
Hmm, why filter out logical partitions? Is there some specification that says that?

poettering replied:
nope, but we are looking for devices that contain GPT partition tables, not the partitions themselves. Or in other words for /dev/sda rather than /dev/sda5 if you so will. Hence we filter out partitions here.

poettering replied:
(will add a comment about that)

keszybz commented on src/boot/efi/util.c:
```diff
@@ -159,8 +162,13 @@ EFI_STATUS efivar_get_raw(const EFI_GUID *vendor, const CHAR16 *name, CHAR8 **bu
+                if (buffer)
+                        *buffer = buf;
+                else
+                        FreePool(buf);
```
We actually have FreePoolp, so
```C
if (buffer)
      *buffer = TAKE_PTR(buf);
```
is possible.

poettering replied:
we have no TAKE_PTR right now in the EFI sources though. will add

keszybz commented on mkosi.build:
```diff
@@ -95,3 +95,7 @@ EOF
+
+mkdir -p "$DESTDIR"/efi/EFI/systemd "$DESTDIR"/efi/EFI/BOOT
+cp "$DESTDIR"/usr/lib/systemd/boot/efi/systemd-bootx64.efi "$DESTDIR"/efi/EFI/systemd/systemd-bootx64.efi
+cp "$DESTDIR"/usr/lib/systemd/boot/efi/systemd-bootx64.efi "$DESTDIR"/efi/EFI/BOOT/bootx64.efi
```
But why would have it in both places? That seems wasteful and overkill for a test image like this.

poettering replied:
Simply because Fedora currently puts it in /boot/efi and we in /efi... This really needs to be fixed but I don't think it's for this PR to fix that.

keszybz commented on man/systemd-gpt-auto-generator.xml:
```diff
@@ -32,22 +32,18 @@
+    url="http://www.uefi.org/specifications">UEFI Specification</ulink>, chapter 5.  It implements the <ulink
+    url="https://www.freedesktop.org/wiki/Specifications/DiscoverablePartitionsSpec/">Discoverable Partitions
+    Specification</ulink>. Note that this generator has no effect on non-GPT systems, or where the
+    directories under the mount points are already non-empty. Also, on systems where the units are explicitly
```
Stuff under `/efi` is ignored, so this sentence is wrong? (Or was it only in dissect-mount?)

poettering replied:
no we do check whether /efi is already populated and if so don't over-mount it. but the text is still a bit skewed, since the generator has effect on other dirs even if some dir is already populated

keszybz commented on man/systemd-gpt-auto-generator.xml:
```diff
@@ -149,44 +162,49 @@
+    device mapper device name.</para>
+
+    <para>Mount and automount units for the EFI System Partition (ESP) are generated on EFI systems. The ESP
+    is mounted to <filename>/boot/</filename>, unless a mount point directory <filename>/efi/</filename>
```
This text should be updated too. Maybe the paragraph about `Extended Boot Loader partition` should be first, and then this one can simply say "ESP is mounted on /boot if it not used for the other, and on /efi otherwise".

keszybz commented on man/systemd-gpt-auto-generator.xml:
```diff
@@ -149,44 +162,49 @@
+    are generated.</para>
+
+    <para>If the disk contains an Extended Boot Loader partition, as defined in the <ulink
+    url="https://systemd.io/BOOT_LOADER_SPECIFICATION">Boot Loader Specification</ulink> it made available on
```
Needs a comma after the ulink.

keszybz commented on man/bootctl.xml:
```diff
@@ -138,9 +154,12 @@
+    relaxed, and the path specified with <option>--esp-path=</option> may refer to any kind of file system on
+    any kind of partition.</para>
+
+    <para>Similar <varname>$SYSTEMD_RELAX_XBOOTLDR_CHECKS=1</varname> turns off some validation checks for
```
Similarly

poettering commented on src/shared/gpt.h:
```diff
@@ -15,6 +15,7 @@
 #define GPT_ROOT_ARM_64 SD_ID128_MAKE(b9,21,b0,45,1d,f0,41,c3,af,44,4c,6f,28,0d,3f,ae)
 #define GPT_ROOT_IA64   SD_ID128_MAKE(99,3d,8d,3d,f8,0e,42,25,85,5a,9d,af,8e,d7,ea,97)
 #define GPT_ESP         SD_ID128_MAKE(c1,2a,73,28,f8,1f,11,d2,ba,4b,00,a0,c9,3e,c9,3b)
+#define GPT_XBOOTLDR    SD_ID128_MAKE(bc,13,c2,ff,59,e6,42,62,a3,52,b2,75,fd,6f,71,72)
```
in the boot loader spec, for a longer time, will add a note to the commit msg

poettering commented on src/shared/bootspec.c:
```diff
@@ -629,6 +629,174 @@ int find_esp_and_warn(
+        if (stat(p, &st) < 0)
+                return log_full_errno((searching && errno == ENOENT) &&
+                                      (unprivileged_mode && errno == EACCES) ? LOG_DEBUG : LOG_ERR, errno,
+                                      "Failed to determine block device node of \"%s\": %m", p);
```
(actually corrected in a later commit already in the same PR, when this was reworked again, but will fix the earlier commit too)

poettering commented:
I force pushed a new version now. I addressed everything, except for the mkosi EFI copy thingy (see above). Which should be fixed, but probably outside of this PR, as it require fedora changes most likely, and possibly kernel-install fixes.

To make review easy, here's a diff between the earlier version and this new version:

https://paste.fedoraproject.org/paste/FttLKUkQScVxdIrV3RxVpg

(Note that the changes in this version are actually bigger, as I fixed some earlier commits properly, that in the old PR version were fixed in later commits).

poettering commented:
Taking liberty to set green label, since the changes are mostly docs, indentation, line break stuff, which should be with little risk. let's speed the process up

bam80 commented:
> Extended Boot Loader partition (herein dubbed "XBOOTLDR") that is defined in https://systemd.io/BOOT_LOADER_SPECIFICATION for a long time

Sorry I didn't find anything about XBOOTLDR in the specification. What happened with it?

keszybz commented:
Yep, it seems the docs are lacking. I filed an issue now so we don't forget.

poettering commented:
The docs describe the partition just fine: it's the one mentioned in the referenced doc with the type uuid  of bc13c2ff-59e6-4262-a352-b275fd6f7172. The spec doesnt specify a name for it though. In code we then call the partition of that type xbootldr. That's all..

Not sure if the spec really needs to call it anything other than by its type uuid.

bam80 commented:
Right now it makes the confusions. I had hard time trying to figure out what XBOOTLDR actually is..
All the others types seem have descriptive names: https://en.wikipedia.org/wiki/GUID_Partition_Table#Partition_type_GUIDs
For examples, both `ESP` and `EFI System partition` names are present in UEFI spec.
I understand that XBOOTLDR is non-standard but still..

bam80 commented:
> > Would the filesystem be any supported by the UEFI system drivers?
> 
> Yes, this code makes no restriction on the file system used (which is different from the handling of the ESP where we explicitly check for FAT). In fact in a future patch I'd like to add a special directory in the ESP to sd-boot where you can drop EFI drivers into, and sd-boot will load them all. The idea is then that people can drop weird fs drivers in there, and sd-boot will load them and then have them available when searching for XBOOTLDR. That way we can support other fs, in a fashion of very lose coupling, as sd-boot needs to know nothing about fs drivers, it just loads everything dropped in.

On practice, it doesn't work for BTRFS, unfortunately. Me and other community members were not able to boot the system from BTRFS XBOOTLDR partition:

> I actually tried that. I marked my root partition as XBOOTLDR and placed a filesystem driver (btrfs_x64.efi) into EFI/systemd/drivers/ of ESP partition.
But unfortunately it does not work. At least with Btrfs that I tried. Systemd-boot does not pick it up for some reason.

[Systemd-boot on btrfs /boot](https://bbs.archlinux.org/viewtopic.php?pid=2130615#p2130615)

Is it a known issue or I should file a bug?
